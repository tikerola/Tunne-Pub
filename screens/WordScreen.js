import React from 'react'
import { StyleSheet, View } from 'react-native'
import { useDispatch, useSelector } from 'react-redux'
import CustomButton from '../components/CustomButton'
import TextSlider from '../components/TextSlider'
import { colors } from '../constants/colors'
import { setScreenNumber, toggleWord } from '../store/actions/cards'

export default function WordScreen({ route, navigation }) {

  const { title, wordType, navigateBackTo, question } = route ? route.params : {}
  
  const words = useSelector(state => wordType === 'sentences' 
  ? 
  state.cards.sentences 
  : 
  (wordType === 'needs' || wordType === 'needsBehindAFeeling' ? state.cards.needs : state.cards.words))

  const dispatch = useDispatch()

  const onSave = () => {
    const screenNumber = (wordType === 'needs' || wordType === 'sentences')
      ?
      3
      :
      (wordType === 'feelingsYouMissWords' ? 2 : 1)

    dispatch(setScreenNumber(screenNumber))
    navigation.navigate(`${navigateBackTo}`)
  }

  return (
    <View style={styles.container}>
      <TextSlider
        textData={words}
        saveToRedux={toggleWord}
        question={question}
        title={title}
        columns={wordType === 'needs' || wordType === 'needsBehindAFeeling' || wordType === 'sentences' ? 1 : 2}
        wordType={wordType}
      />
      <CustomButton title="Tallenna" onPress={onSave} />

    </View>

  )
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    backgroundColor: colors.bodyBg
  }
})
