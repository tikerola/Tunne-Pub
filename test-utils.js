import React from "react";
import { NavigationContext } from "@react-navigation/native";
import { render } from "@testing-library/react-native";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import CARD_DATA from "./data/cards";
import NEEDS_DATA from "./data/needs";
import { SENTENCES_DATA } from "./data/sentences";
import WORD_DATA from "./data/words";

const INITIAL_STATE = {
  cards: {
    // cards: CARD_DATA,
    // words: WORD_DATA,
    // needs: NEEDS_DATA,
    // sentences: SENTENCES_DATA,
    activeSummaryItemId: "123",
    summaryOfChosenItems: {
      123: {
        id: "123",
        name: "",
        cards: [],
        feelingsWords: [],
        feelingsYouMissWords: [],
        needs: ["hei"],
        needsBehindAFeeling: [],
        sentences: [],
        questions: {},
        itemSaved: true,
      },
    },
    screenNumber: 1,
    numOfItems: 0,
    initialLoadFinished: false,
  },
};

const mockStore = configureStore([]);
const store = mockStore(INITIAL_STATE);

const navContext = {
  isFocused: () => true,
  // addListener returns an unscubscribe function.
  addListener: jest.fn(() => jest.fn()),
};

const reduxRender = (ui, options, storeData = store) =>
  render(ui, {
    wrapper: ({ children }) => (
      <Provider store={storeData}>
        <NavigationContext.Provider value={navContext}>
          {children}
        </NavigationContext.Provider>
      </Provider>
    ),
    ...options,
  });

// re-export everything for convenience
export * from "@testing-library/react-native";

// override render method
export { reduxRender as render, store };
