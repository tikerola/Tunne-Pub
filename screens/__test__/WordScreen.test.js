import { waitFor } from '@testing-library/react-native'
import React from 'react'
import { render, fireEvent, store } from '../../test-utils'
import WordScreen from '../WordScreen'

afterAll(() => {
    store.clearActions()
})

describe('<WordScreen />', () => {

    const navigationMock = {
        navigate: jest.fn()
    }

    it('should match snapshot', () => {
        const result = render(<WordScreen />).toJSON()

        expect(result).toMatchSnapshot()
    })

    it('should handle onSave, when wordType is feelingsWords', async () => {

        const mockRoute = {
            params: { wordType: 'feelingsWords', navigateBackTo: 'MainScreen'}
        }

        const rendered = render(<WordScreen navigation={navigationMock} route={mockRoute} />)
        const button = rendered.getByTestId('button')

        fireEvent.press(button)

        const actions = await waitFor(() => store.getActions())
        
        expect(actions[0].type).toEqual('SET SCREEN NUMBER')
        expect(actions[0].screenNumber).toBe(1)
        expect(navigationMock.navigate.mock.calls[0][0]).toEqual('MainScreen')

    })

    it('should handle onSave, when wordType is needs', async () => {

        store.clearActions()

        const mockRoute = {
            params: { wordType: 'needs', navigateBackTo: 'ThirdScreen'}
        }

        const rendered = render(<WordScreen navigation={navigationMock} route={mockRoute} />)
        const button = rendered.getByTestId('button')

        fireEvent.press(button)

        const actions = await waitFor(() => store.getActions())
        
        expect(actions[0].type).toEqual('SET SCREEN NUMBER')
       
        expect(actions[0].screenNumber).toBe(3)
        expect(navigationMock.navigate).toHaveBeenCalledWith('ThirdScreen')

    })
})