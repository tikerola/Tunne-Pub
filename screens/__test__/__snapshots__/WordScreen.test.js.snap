// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<WordScreen /> should match snapshot 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#5f0c38",
      "flex": 1,
    }
  }
>
  <RCTSafeAreaView
    emulateUnlessSupported={true}
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flexDirection": "column",
        "justifyContent": "center",
        "width": "100%",
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#999",
          "fontFamily": "Roboto-Light",
          "fontSize": 16,
          "marginBottom": 10,
          "marginLeft": "15%",
          "marginTop": 10,
          "width": "100%",
        }
      }
    >
      *
      Sanalista
       jatkuu vierittämällä näyttöä
    </Text>
    <RCTScrollView
      disableVirtualization={false}
      getItem={[Function]}
      getItemCount={[Function]}
      horizontal={false}
      initialNumToRender={20}
      keyExtractor={[Function]}
      maxToRenderPerBatch={10}
      numColumns={2}
      onContentSizeChange={[Function]}
      onEndReachedThreshold={2}
      onLayout={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={Array []}
      style={
        Object {
          "height": 733.7,
          "marginBottom": 15,
          "marginTop": 10,
          "width": "95%",
        }
      }
      testID="flatlist"
      updateCellsBatchingPeriod={50}
      viewabilityConfigCallbackPairs={Array []}
      windowSize={21}
    >
      <View />
    </RCTScrollView>
  </RCTSafeAreaView>
  <View
    accessible={true}
    focusable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignSelf": "center",
        "backgroundColor": "#3a89ed",
        "borderRadius": 5,
        "elevation": 5,
        "margin": 0,
        "opacity": 1,
        "padding": 5,
        "shadowColor": "#000",
        "shadowOffset": Object {
          "height": 4,
          "width": 0,
        },
        "shadowOpacity": 0.44,
        "shadowRadius": 5,
        "width": 100,
      }
    }
    testID="button"
  >
    <Text
      style={
        Object {
          "alignSelf": "center",
          "color": "white",
          "fontFamily": "Roboto-Light",
          "fontSize": 16,
        }
      }
      testID="title"
    >
      Tallenna
    </Text>
  </View>
</View>
`;
